spec.html

```diff
@@ -7413,9 +7413,8 @@ <h1>FunctionAllocate ( _functionPrototype_, _strict_, _functionKind_ )</h1>
       <h1>FunctionInitialize ( _F_, _kind_, _ParameterList_, _Body_, _Scope_ )</h1>
       <p>The abstract operation FunctionInitialize requires the arguments: a function object _F_, _kind_ which is one of (Normal, Method, Arrow), a parameter list Parse Node specified by _ParameterList_, a body Parse Node specified by _Body_, a Lexical Environment specified by _Scope_. FunctionInitialize performs the following steps:</p>
       <emu-alg>
-        1. Assert: _F_ is an extensible object that does not have a `length` own property.
         1. Let _len_ be the ExpectedArgumentCount of _ParameterList_.
-        1. Perform ! DefinePropertyOrThrow(_F_, `"length"`, PropertyDescriptor{[[Value]]: _len_, [[Writable]]: *false*, [[Enumerable]]: *false*, [[Configurable]]: *true*}).
+        1. Perform ! SetFunctionLength(_F_, _len_).
```

Comment:
Not sure whether this should be prefixed with "!". I assume it does, but then SetFunctionName often isn't elsewhere in the spec and I'm not sure why.

Reply:
If an operation cannot *ever* return an abrupt completion, then I'd say calls to it shouldn't be preceded by '!'. It isn't *wrong*, but it's unnecessary clutter. (This would be more persuasive if you could easily tell from an operation's 'declaration' whether it ever returns an abrupt completion. See issue #253.)

Reply:
So just to clarify my understanding here:

While `DefinePropertyOrThrow` can throw in some circumstances, it can't in this case, because we know from the first assertion in `SetFunctionLength` that `DefinePropertyOrThrow`'s first argument is extensible and doesn't have a `length` own prop which could interfere with setting.

Thus we know that `SetFunctionLength` won't ever throw.

Preceding an abstract operation call by "!" means that we are in a situation where an abstract operation won't throw, _although it might do so in different circumstances_.

Operations which never throw should never be prefixed with "!"  even though we can't know from looking at the abstract operation's signature or its name that this is the case.

So for now, `SetFunctionLength` should not be prefixed with anything?

Is my understanding correct?

Reply:
Your understanding matches mine. Whether it's *correct* is up to the editors.

Reply:
I don't agree with this; I think we should use ! everywhere. (I am also not an editor.)